packages/core/src/internal/fetchOptions.ts

```diff
@@ -129,9 +129,28 @@ export const makeFetchOptions = (
     (typeof operation.context.fetchOptions === 'function'
       ? operation.context.fetchOptions()
       : operation.context.fetchOptions) || {};
-  if (extraOptions.headers)
-    for (const key in extraOptions.headers)
-      headers[key.toLowerCase()] = extraOptions.headers[key];
+  if (extraOptions.headers) {
+    if (extraOptions.headers.forEach) {
+      (extraOptions.headers as Headers | [[string, string]]).forEach(
+        (value, key) => {
```

Comment:
I'm a bit concerned that the `forEach`’s function doesn't infer from usage (i.e. `value` and `key` are `any`)

Also `[[string, string]]` is incorrect, as it's `[string, string][]`, but this doesn't fix the inference here anyway, so it's a bit besides the point.

The main problem here is that both cases aren't equivalent however. `forEach` on the array of tuples will have `value` be `[string, string]`, while on `Headers#forEach` it'll be just `string` and the second argument becomes `string` too.

We can supplement this with `headers.entries()` but that invalidates the type cast.
It also is a little awkward since we'd have to check `Array.isArray` and otherwise check for something that definitively tells us we have a `Headers` instance.
We can still rely on `forEach`, but we could also do `'has' in headers && !Object.keys(headers).length` which is pretty definitive.

But this comes down to `for (const entry of Array.isArray(headers) ? headers : (headers as Headers).entries())`, which will also require `dom.iterable` to be added to our `tsconfig.json`.

---

Another alternative is to reuse whatever we receive:

```ts
  const defaultHeaders = {
    accept:
      operation.kind === 'subscription'
        ? 'text/event-stream, multipart/mixed'
        : 'application/graphql-response+json, application/graphql+json, application/json, text/event-stream, multipart/mixed',
  };

  const extraOptions =
    (typeof operation.context.fetchOptions === 'function'
      ? operation.context.fetchOptions()
      : operation.context.fetchOptions) || {};

  const isHeaders = (headers: HeadersInit): headers is Headers =>
    'has' in headers && !Object.keys(headers).length;

  let headers: HeadersInit | undefined = extraOptions.headers;
  if (!headers) {
    headers = defaultHeaders;
  } else if (Array.isArray(headers)) {
    if (!headers.some((entry) => entry[0].toLowerCase() === 'accept'))
      headers.push(['accept', defaultHeaders.accept]);
  } else if (isHeaders(headers)) {
    if (!headers.has('accept'))
      headers.set('accept', defaultHeaders.accept);
  } else {
    if (!Object.keys(headers).some((key) => key.toLowerCase() === 'accept'))
      headers.accept = defaultHeaders.accept;
  }
```

which also avoids the discrepancy below of `${headerPrevValue},${headerValue}` rather than `${headerPrevValue}, ${headerValue}` and any similar issues we may run into.

Ultimately either is fine though 🤔